src/manifest_store.rs

```diff
@@ -161,8 +180,77 @@ impl StoredManifest {
         Ok(&self.manifest.core)
     }
 
+    /// Create a new checkpoint from the current state. This only creates the checkpoint
+    /// struct, but does not persist it in the manifest. The latter can be done with
+    /// [`Self::write_checkpoint`].
+    pub(crate) fn new_checkpoint(
+        &self,
+        checkpoint_id: Uuid,
+        options: &CheckpointOptions,
+    ) -> Result<Checkpoint, SlateDBError> {
+        let expire_time = options.lifetime.map(|l| SystemTime::now() + l);
+        let db_state = self.db_state();
+        let manifest_id = match options.source {
+            Some(source_checkpoint_id) => {
+                let Some(source_checkpoint) = db_state.find_checkpoint(&source_checkpoint_id)
+                else {
+                    return Err(InvalidDBState);
+                };
+                source_checkpoint.manifest_id
+            }
+            None => {
+                if !db_state.initialized {
+                    return Err(InvalidDBState);
+                }
+                self.id()
+            }
+        };
+        Ok(Checkpoint {
+            id: checkpoint_id,
+            manifest_id,
+            expire_time,
+            create_time: SystemTime::now(),
+        })
+    }
+
+    /// Write the given checkpoint to the manifest,
+    pub(crate) async fn write_checkpoint(
+        &mut self,
+        checkpoint: Checkpoint,
+    ) -> Result<(), SlateDBError> {
+        self.maybe_apply_db_state_update(|stored_manifest| {
+            // TODO: Do we need to verify that the manifest is still available
+            //       before writing?
+            let mut updated_db_state = stored_manifest.db_state().clone();
+            updated_db_state.checkpoints.push(checkpoint.clone());
```

Comment:
I think we should do a safety check to make sure the checkpoint isn't already in the db state.